### PR TITLE
net: pkt: Compute TX payload data length 

### PR DESCRIPTION
**Pull request by pfl**
Compute the length of the TX payload that is transported in one IPv4 or IPv6 datagram taking into account UDP, ICMP or TCP headers in addition to any IPv6 extension headers added by RPL. The TCP implementation in Zephyr is known to currently carry at maximum 8 bytes of options. If the protocol is not known to the stack, assume that the application handles any protocol headers as well as the data. Also, if the net_pkt does not have a context associated, length check on the data is omitted when appending.
    
Although payload length is calculated also for TCP, the TCP MSS value is used as before.
    
Define IPv4 minimum MTU as 576 octets, See RFC 791, Sections 3.1. and 3.2.
    
Signed-off-by: Patrik Flykt <patrik.flykt@intel.com>

### CODE REVIEW

**Comment by pfl** [comment]
This is definitely a Request For Comments patch, let's continue coding on monday when everybody is back. The IPv6 upper layer header should be stored in advance, i.e. when the extension header lengths are computed.

**Comment by pfl** [comment]
Need to compute header lengths already at packet allocation times, updated RFC patch accordingly.

File: include/net/net_pkt.h
Comment (tbursztyka) [inline include/net/net_pkt.h]: I think this could be avoided, see below

File: subsys/net/ip/net_pkt.c
Comment (tbursztyka) [inline subsys/net/ip/net_pkt.c]: all the code block here is not that much time/performance consuming so (see below)

File: subsys/net/ip/net_pkt.c
Comment (tbursztyka) [inline subsys/net/ip/net_pkt.c]: it could anyway be called here everytime and available space would be computed from that minus current total length.

no?

**Review by tbursztyka — COMMENTED** [review]
Maybe we should try to avoid adding another attribute to net_pkt

File: subsys/net/ip/net_pkt.c
Comment (pfl) [reply subsys/net/ip/net_pkt.c]: We need to remember the amount of space remaining so that two consecutive calls to net_pkt_append() succeed, unless there is another application data counter somewhere in the struct already?

Comment (jukkar) [reply subsys/net/ip/net_pkt.c]: As @pfl mentioned, we need to support the case where multiple append/insert calls are done, so the amount of bytes needs to be remembered.

Comment (tbursztyka) [reply subsys/net/ip/net_pkt.c]: But we know that: we can get the actual occupied length in the chain of net_buf.

**Comment by jukkar** [comment]
Briefly tested this with echo-server and http-server running in zephyr side. Seemed to work just fine. Will test more tomorrow.

File: subsys/net/ip/net_pkt.c
Comment (pfl) [reply subsys/net/ip/net_pkt.c]: What variables would we need to look at?

Comment (tbursztyka) [reply subsys/net/ip/net_pkt.c]: I though the whole length of fragment list, net_buf_frags_len(pkt->frags), could help. But as the ll reserve part is badly managed for now, it won't work.

**Comment by pfl** [comment]
Coap tests succeed with the latest update.

File: subsys/net/ip/net_pkt.c
Comment (rveerama1) [inline subsys/net/ip/net_pkt.c]: How about considering net_pkt_set_ipv6_ext_len() in case ipv6 and move away RPL check here. RPL setting this value when insert is success "net_pkt_set_ipv6_ext_len(pkt, ext_len + NET_RPL_HOP_BY_HOP_LEN);".

File: include/net/net_pkt.h
Comment (pfalcon) [reply include/net/net_pkt.h]: +1 . From my comment https://github.com/zephyrproject-rtos/zephyr/pull/119#issuecomment-339361679 :

> As far as I can tell, for TX packet, net_pkt->appdatalen isn't used until packet construction is finished and ready to be sent. So, it can be used to cache "remaining data size in the packet" (initialized in net_pkt_get_tx()).

**Comment by pfalcon** [comment]
Inspired by @jukkar's comment at https://github.com/zephyrproject-rtos/zephyr/pull/5015#issuecomment-346007196 :

> It looks like #5038 could replace this (#5015) one, also then we would not need to have #5034 either.

I definitely appreciate this patch, the questions is its scope: #5015/#5034 were proposed as safe (IMHO) and acceptable workarounds to get the 1.10 out. This patch is a bigger changes in itself, and based on the comments, there're various things which could be tweaked/improved. Up to @jukkar / @tbursztyka then whether this could fit into 1.10. (A possible compromise might be merge mostly as is as long as it works, and leave further refactors/optimizations until new merge window opens).

File: subsys/net/ip/net_pkt.c
Comment (pfl) [reply subsys/net/ip/net_pkt.c]: RPL has not yet set the ext header length when this function is called.

**Comment by pfl** [comment]
Commit message and its header updated, unnecessary setting variable value on initialization removed.

**Review by jukkar — APPROVED** [review]
I have been testing this and everything has been working just fine so +1 from me.

**Comment by pfalcon** [comment]
Ok, if this is considered for merging, I'd like to ask to update the commit message to explicitly state that the change is for TX packets. E.g., the title can be (at the very least) `net: pkt: Compute TX payload data length` (better `net: pkt: Compute payload data length for TX packets`), and corresponding additions in the commit description body.

**Comment by pfalcon** [comment]
More of commit message:

> TCP is known to carry at maximum 8 bytes of options.

Well, that's to broad a statement to make. The *TCP* (RFC 793) definitely can carry more than 8 bytes of options. We have the NET_TCP_MAX_OPT_SIZE define with a random value, apparently inherited from FNET, removing of which (FNET cruft) is on my TODO.


So, suggestion: keep the code as is (as it's just first iteration of the matter, will be refactoring/optimizing it later based on all suggestions), but remove that sentence from the commit message.

I'd even ask why all that is at all relevant for the TCP case, if

> TCP payload length is not changed, TCP MSS is used as before.

But don't let me go that rabbit hole, again, pretty good for the first iteration.

**Review by pfalcon — CHANGES_REQUESTED** [review]
I'd really appreciate the commit message changes as suggested above, to not create unneeded confusion. Thanks.

**Comment by pfl** [comment]
> Well, that's to broad a statement to make. The TCP (RFC 793) definitely can carry more than 8 bytes of options.

*This* patch relates to *Zephyr's* implementation of TCP, so that is an accurate statement to make, actually.

> I'd even ask why all that is at all relevant for the TCP case, if

It is relevant, since it computes lengths already for the TCP case, even though tcp->mss is still being used. That again is for another patch, but not for 1.10, since no release critical bugs are reported for TCP right now. TCP works fine, AFAIK by glancing at Jukka's screen where he's sending and receiving something continuosly.

**Comment by pfalcon** [comment]
> This patch relates to Zephyr's implementation of TCP, so that is an accurate statement to make, actually.

As an outside reader, I find the phrase "TCP is known to carry at maximum 8 bytes of options." to be highly ambiguous/confusing, that's why I ask that to be dealt with.

**Comment by pfl** [comment]
Updated commit message.

**Review by pfalcon — APPROVED** [review]
Thanks!

**Comment by pfl** [comment]
AFAIK, together with the already merged #5072, this PR fixes #4934, and obsoletes #5034, #4972 and #5015.

**Review by tbursztyka — APPROVED** [review]

**Comment by pfalcon** [comment]
Thanks for merging.

> AFAIK, together with the already merged #5072, this PR fixes #4934, and obsoletes #5034, #4972 and #5015.

@jukkar closed #4972, and I closed #5015 & #5034.

**Pull request merged**
